cpp/include/cuvs/neighbors/ann_mg.hpp

```diff
@@ -0,0 +1,1367 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#ifdef CUVS_BUILD_MG_ALGOS
+
+#include <raft/core/device_resources.hpp>
+#include <rmm/mr/device/per_device_resource.hpp>
+
+#include <cuvs/neighbors/cagra.hpp>
+#include <cuvs/neighbors/ivf_flat.hpp>
+#include <cuvs/neighbors/ivf_pq.hpp>
+
+#ifndef NO_NCCL_FORWARD_DECLARATION
+class ncclComm_t {};
+#endif
+
+#define DEFAULT_SEARCH_BATCH_SIZE 1 << 20
+
+/// \defgroup ann_mg_cpp_index_params ANN MG index build parameters
+
+namespace cuvs::neighbors::mg {
+/** Distribution mode */
+/// \ingroup ann_mg_cpp_index_params
+enum distribution_mode {
+  /** Index is replicated on each device, favors throughput */
+  REPLICATED,
+  /** Index is split on several devices, favors scaling */
+  SHARDED
+};
+}  // namespace cuvs::neighbors::mg
+
+namespace cuvs::neighbors::ivf_flat {
+/// \ingroup ann_mg_cpp_index_params
+/**
+ * IVF-Flat MG build parameters struct
+ *
+ * Usage example:
+ * @code{.cpp}
+ * cuvs::neighbors::ivf_flat::mg_index_params index_params;
+ * index_params.mode = cuvs::neighbors::mg::distribution_mode::SHARDED;
+ * index_params.metric = cuvs::distance::DistanceType::L2Expanded;
+ * index_params.metric_arg = 2.0f;
+ * index_params.add_data_on_build = true;
+ * @endcode
+ *
+ */
+struct mg_index_params : public cuvs::neighbors::ivf_flat::index_params {
+  /** Distribution mode */
+  cuvs::neighbors::mg::distribution_mode mode;
+};
+}  // namespace cuvs::neighbors::ivf_flat
```

Comment:
What do you think about generalizing the mg_index_params into something like the following?
```C++
namespace cuvs::neighbors::mg {
template <typename Upstream>
struct mg_params :: public Upstream {
  /** Distribution mode */
  distribution_mode mode;
};
}
```
Then, you could use this in the search/build functions directly, with two benefits:
  1. No need to create a new type for every existing algorithm
  2. Potentially, you could make some of the functions in the detail namespace use the templated version of the struct rather than duplicating code
```C++
template <typename T, typename IdxT>
void search(const raft::resources& handle,
            const cuvs::neighbors::mg::nccl_clique& clique,
            const ann_mg_index<ivf_flat::index<T, IdxT>, T, IdxT>& index,
            const mg_params<ivf_flat::search_params>& search_params,
            raft::host_matrix_view<const T, IdxT, row_major> queries,
            raft::host_matrix_view<IdxT, IdxT, row_major> neighbors,
            raft::host_matrix_view<float, IdxT, row_major> distances)
```
(In this example, one could also modify ivf_flat  `index` type to add `using search_params_type = ivf_flat::search_params`, and then the whole MG search could be templated over the index type (i.e. `mg_params<typename Upstream::search_params_type>`))

Reply:
Here's an example of how I get around this with a similar approach in a dynamic batching prototype (but the API is surely simpler than this multi-gpu PR, because I only need the search methods) https://github.com/rapidsai/cuvs/pull/261/files#diff-89b13a31203fa009604a3a2bdb0aee01668bb68969d9193a2b31d6bf6dd3894aR29-R48

Reply:
I like the idea of improving the API through templating (or even dynamic batching) and keeping everything under the `cuvs::neighbors::mg` namespace. The `distribution_mode` is only necessary during build time and the `merge_mode` during search time though. The `merge_mode` could be incorporated in the search parameter. However, there's way too many things to consider to modify this in time for the release. I think we could allow ourselves a one-time update of the public API if necessary during the second release as not too many people (we could hope) would be using it yet.